src/libopensc/card-nqApplet.c

```
@@ -0,0 +1,556 @@
+{
+	struct sc_apdu apdu;
+	int rv;
+	size_t resplen;
+
+	LOG_FUNC_CALLED(card->ctx);
+	resplen = MIN(sc_get_max_recv_size(card), *cb_resp);
+
+	sc_format_apdu_ex(&apdu, 0x80, 0xC0, 0x00, 0x00, NULL, 0, resp, resplen);
+	apdu.flags  |= SC_APDU_FLAGS_NO_GET_RESP;
+
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, rv, "APDU transmit failed");
+	if (apdu.resplen == 0)
+	{
+		LOG_FUNC_RETURN(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2));
+	}
+
+	*cb_resp = apdu.resplen;
+
+	if (apdu.sw1 == 0x90 && apdu.sw2 == 0x00)
+	{
+		rv = SC_SUCCESS;
+	}
+	else if (apdu.sw1 == 0x61)
+	{
+		rv = apdu.sw2 == 0 ? 256 : apdu.sw2;
+	}
+	else if (apdu.sw1 == 0x62 && apdu.sw2 == 0x82)
+	{
+		rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	}
+
+	LOG_FUNC_RETURN(card->ctx, rv);
+}
+
+static int nqapplet_get_challenge(struct sc_card *card, u8 * buf, size_t count)
+{
+	int r;
+	struct sc_apdu apdu;
+
+	LOG_FUNC_CALLED(card->ctx);
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2, 0x84, 0x00, 0x00);
+	apdu.le = count;
+	apdu.resp = buf;
+	apdu.resplen = count;
+
+	r = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
+
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(card->ctx, r, "GET CHALLENGE failed");
+
+	if (count < apdu.resplen)
+	{
+		return (int) count;
+	}
+   
+	return (int) apdu.resplen;
+}
+
+static int nqapplet_logout(struct sc_card *card)
+{
+	LOG_FUNC_CALLED(card->ctx);
+
+	int rv = select_nqapplet(card, NULL, NULL, NULL, 0, NULL);
+	if ( rv != SC_SUCCESS )
+	{
+		LOG_TEST_RET(card->ctx, rv, "Failed to logout");
+	}
+	
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+int nqapplet_set_security_env(struct sc_card *card, const struct sc_security_env *env, int se_num)
+{
+	/* Note: the NQ-Applet does not have APDU for SET SECURITY ENV, 
+	this function checks the intended parameters and sets card_data.key_reference */
+	nqapplet_driver_data_ptr data;
+	
+	assert(card != NULL && env != NULL);
```

please return an error code instead

---

It seems that other card drivers don't bother with parameter checking, so I can assume that at this point these parameters are correct. Removed the asserts.